passes/hierarchy/hierarchy.cc

```diff
@@ -808,6 +808,37 @@ struct HierarchyPass : public Pass {
 				if (mod_it.second->get_bool_attribute("\\top"))
 					top_mod = mod_it.second;
 
+		if (top_mod != nullptr && auto_top_mode) {
```

Comment:
I don't think this is interacting with `-auto-top` correctly.

The idea of `-auto-top` is that it finds a suitable top module when the design contains no marked top module. But this case is one where in fact the design contains a marked top-module. This would derive from a marked abstract top module when `-auto-top` is used, but not derive from it otherwise?!

It inherently doesn't make any sense imo to combine `-auto-top` with `-chparam`, but this seems to actually require to combine the two.

I think the following is the intended behavior.

```
                if (top_mod != nullptr && top_mod->name.begins_with("$abstract"))
                {
                        IdString top_name = top_mod->name.substr(strlen("$abstract"));

                        dict<RTLIL::IdString, RTLIL::Const> top_parameters;
                        for (auto &para : parameters) {
                                SigSpec sig_value;
                                if (!RTLIL::SigSpec::parse(sig_value, NULL, para.second))
                                        log_cmd_error("Can't decode value '%s'!\n", para.second.c_str());
                                top_parameters[RTLIL::escape_id(para.first)] = sig_value.as_const();
                        }

                        top_mod = design->module(top_mod->derive(design, top_parameters));

                        if (top_mod != nullptr && top_mod->name != top_name) {
                                Module *m = top_mod->clone();
                                m->name = top_name;
                                Module *old_mod = design->module(top_name);
                                if (old_mod)
                                        design->remove(old_mod);
                                design->add(m);
                                top_mod = m;
                        }
                }
```

Reply:
>  This would derive from a marked abstract top module when -auto-top is used, but not derive from it otherwise?!

No, the intention is that -- with `(* top *)` present, it would behave the same with and without `-auto-top` since `synth_*` calls `hierarchy -auto-top` in the absence of a specific `-top`.

I see that my PR currently doesn't do that... let me fix it.

Reply:
> It inherently doesn't make any sense imo to combine -auto-top with -chparam, but this seems to actually require to combine the two.

I can't see any reason why this wouldn't be desirable?

Remember, ultimately I want to see `read_verilog -defer` behave the same as `read_verilog` (equally, `read` before and after #1233) when there is a `(* top *)` involved, including with `-chparam`.